ibtk/src/math/PETScMatUtilities.cpp

```diff
@@ -806,10 +618,460 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
 
 /////////////////////////////// PRIVATE //////////////////////////////////////
 
+void PETScMatUtilities::constructPatchLevelProlongationOp_cell(Mat& mat,
+                                                               int dof_index_idx,
+                                                               const std::vector<int>& num_fine_dofs_per_proc,
+                                                               const std::vector<int>& num_coarse_dofs_per_proc,
+                                                               Pointer<PatchLevel<NDIM> > fine_patch_level,
+                                                               Pointer<PatchLevel<NDIM> > coarse_patch_level,
+                                                               const AO& coarse_level_ao,
+                                                               const int coarse_ao_offset)
+{
+    int ierr;
+    if (mat)
+    {
+        ierr = MatDestroy(&mat);
+        IBTK_CHKERRQ(ierr);
+    }
+
+    // Determine the grid and data extents for the coarse level.
+    const BoxArray<NDIM>& coarse_domain_boxes = coarse_patch_level->getPhysicalDomain();
+#if !defined(NDEBUG)
+    TBOX_ASSERT(coarse_domain_boxes.size() == 1);
+#endif
+    const Index<NDIM>& coarse_domain_lower = coarse_domain_boxes[0].lower();
+    const Index<NDIM>& coarse_domain_upper = coarse_domain_boxes[0].upper();
+    Index<NDIM> coarse_num_cells = 1;
+    coarse_num_cells += coarse_domain_upper - coarse_domain_lower;
+
+    // Ratio between fine and coarse levels.
+    const IntVector<NDIM>& coarse_ratio = coarse_patch_level->getRatio();
+    const IntVector<NDIM>& fine_ratio = fine_patch_level->getRatio();
+    const IntVector<NDIM> fine_coarse_ratio = fine_ratio / coarse_ratio;
+
+    // Determine the matrix dimensions and index ranges.
+    const int mpi_rank = SAMRAI_MPI::getRank();
+    const int m_local = num_fine_dofs_per_proc[mpi_rank];
+    const int n_local = num_coarse_dofs_per_proc[mpi_rank];
+    const int i_fine_lower =
+        std::accumulate(num_fine_dofs_per_proc.begin(), num_fine_dofs_per_proc.begin() + mpi_rank, 0);
+    const int i_fine_upper = i_fine_lower + m_local;
+    const int j_coarse_lower =
+        std::accumulate(num_coarse_dofs_per_proc.begin(), num_coarse_dofs_per_proc.begin() + mpi_rank, 0);
+    const int j_coarse_upper = j_coarse_lower + n_local;
+
+    // Determine the non-zero matrix structure for constant refine.
+    std::vector<int> d_nnz(m_local, 0), o_nnz(m_local, 0);
+    for (PatchLevel<NDIM>::Iterator p(fine_patch_level); p; p++)
+    {
+        Pointer<Patch<NDIM> > fine_patch = fine_patch_level->getPatch(p());
+        const Box<NDIM>& fine_patch_box = fine_patch->getBox();
+        Pointer<CellData<NDIM, int> > dof_fine_data = fine_patch->getPatchData(dof_index_idx);
+        const unsigned depth = dof_fine_data->getDepth();
+        std::vector<int> samrai_petsc_map(depth), local_row(depth);
```

Comment:
Similar comment as in the interface --- use map for an actual mapping. These should be something like "idx", shouldn't they, or are they really "maps"?